src/util/sitemapper.ts

```diff
@@ -0,0 +1,294 @@
+import { Readable } from "stream";
+import { ReadableStream } from "node:stream/web";
+import EventEmitter from "events";
+
+import sax from "sax";
+import PQueue from "p-queue";
+
+import { logger, formatErr } from "./logger.js";
+
+const SITEMAP_CONCURRENCY = 5;
+
+export type Counter = {
+  value: number;
+};
+
+export type SitemapOpts = {
+  headers?: Record<string, string>;
+  q?: PQueue;
+  seenSitemapSet?: Set<string>;
+
+  fromDate?: Date;
+  toDate?: Date;
+  counter?: Counter;
+  limit?: number;
+};
+
+export class SitemapReader extends EventEmitter {
+  headers?: Record<string, string>;
+
+  fromDate?: Date;
+  toDate?: Date;
+
+  q: PQueue;
```

Comment:
Might be slightly better practice to name this `queue`? The type annotations really help clarify intention though, so feel free to ignore.

Reply:
done!